app/src/main/java/io/gnosis/safe/ui/settings/app/passcode/DisablePasscodeFragment.kt

```diff
@@ -0,0 +1,109 @@
+import android.view.ViewGroup
+import android.view.inputmethod.EditorInfo
+import androidx.core.content.ContextCompat
+import androidx.core.widget.doOnTextChanged
+import androidx.navigation.fragment.findNavController
+import io.gnosis.data.security.HeimdallEncryptionManager
+import io.gnosis.safe.R
+import io.gnosis.safe.ScreenId
+import io.gnosis.safe.databinding.FragmentPasscodeBinding
+import io.gnosis.safe.di.components.ViewComponent
+import io.gnosis.safe.ui.base.SafeOverviewBaseFragment
+import io.gnosis.safe.ui.base.fragment.BaseViewBindingFragment
+import io.gnosis.safe.ui.settings.app.SettingsHandler
+import pm.gnosis.svalinn.common.utils.showKeyboardForView
+import pm.gnosis.svalinn.common.utils.snackbar
+import pm.gnosis.svalinn.common.utils.visible
+import javax.inject.Inject
+
+class DisablePasscodeFragment : BaseViewBindingFragment<FragmentPasscodeBinding>() {
+
+    override fun screenId() = ScreenId.SETTINGS_APP_PASSCODE
+
+    @Inject
+    lateinit var settingsHandler: SettingsHandler
+
+    @Inject
+    lateinit var encryptionManager: HeimdallEncryptionManager
+
+
+    override fun inflateBinding(inflater: LayoutInflater, container: ViewGroup?): FragmentPasscodeBinding =
+        FragmentPasscodeBinding.inflate(inflater, container, false)
+
+    override fun inject(component: ViewComponent) {
+        component.inject(this)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        binding.input.showKeyboardForView()
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        with(binding) {
+            title.setText(R.string.settings_passcode_enter_passcode)
+
+            status.visibility = View.INVISIBLE
+
+            createPasscode.setText(R.string.settings_passcode_enter_your_current_passcode)
+            backButton.setOnClickListener {
+                findNavController().popBackStack(R.id.disablePasscodeFragment, true)
+            }
+
+            val digits = listOf(digit1, digit2, digit3, digit4, digit5, digit6)
+            input.showKeyboardForView()
+
+            //Disable done button
+            input.setOnEditorActionListener { _, actionId, _ ->
+                actionId == EditorInfo.IME_ACTION_DONE
+            }
+
+            input.doOnTextChanged { text, _, _, _ ->
+                text?.let {
+                    if (input.text.length < 6) {
+                        digits.forEach {
+                            it.background = ContextCompat.getDrawable(requireContext(), R.color.surface_01)
+                        }
+                        (1..text.length).forEach { i ->
+                            digits[i - 1].background = ContextCompat.getDrawable(requireContext(), R.drawable.ic_circle_passcode_filled_20dp)
+                        }
+                    } else {
+
+                        val success = encryptionManager.unlockWithPassword(text.toString().toByteArray())
```

Comment:
this should be moved to a view model and tested

Reply:
Moved to viewModel. Not yet tested.